twake/backend/node/src/services/notifications/services/mobile-push/service.ts

```diff
@@ -0,0 +1,78 @@
+import Repository from "../../../../core/platform/services/database/services/orm/repository/repository";
+import { DatabaseServiceAPI } from "../../../../core/platform/services/database/api";
+import UserDevice, { TYPE as UserDeviceType } from "../../../../services/user/entities/user_device";
+import { NotificationConfiguration, PushNotificationMessage } from "../../types";
+import { logger } from "../../../../core/platform/framework/logger";
+
+export class MobilePushService {
+  name: "MobilePushService";
+  version: "1";
+  repository: Repository<UserDevice>;
+
+  constructor(
+    private database: DatabaseServiceAPI,
+    private pushConfiguration: NotificationConfiguration["push"],
+  ) {}
+
+  async init(): Promise<this> {
+    this.repository = await this.database.getRepository<UserDevice>(UserDeviceType, UserDevice);
+    return this;
+  }
+
+  //Fixme: add a bulk system to group requests to fcm
+  async push(message: PushNotificationMessage): Promise<void> {
+    if (this.pushConfiguration.type !== "fcm") {
+      logger.info(`${this.name} - Push configuration not set to fcm`);
+      return;
+    }
+
+    const firebaseEndpoint = this.pushConfiguration.fcm.endpoint;
+    const firebaseApiKey = this.pushConfiguration.fcm.key;
+
+    // Get devices and loop over devices
+    const userId = message.user;
+    (await this.repository.find({ user_id: userId })).getEntities().forEach(async device => {
+      const deviceIdentifier = device.value;
+
+      const deviceData = {
+        company_id: message.company_id,
+        workspace_id: message.workspace_id,
+        channel_id: message.channel_id,
+        message_id: message.message_id,
+        thread_id: message.thread_id,
+      };
+
+      const pushMessage = {
+        data: {
+          notification_data: deviceData,
+          click_action: "FLUTTER_NOTIFICATION_CLICK",
+        },
+        notification: {
+          title: message.title,
+          body: message.text,
+          sound: "default",
+          badge: message.badge_value,
+          click_action: "FLUTTER_NOTIFICATION_CLICK",
+        },
+        collapse_key: message.channel_id,
+        registration_ids: [deviceIdentifier],
+      };
+
+      //Push to fcm
+      try {
+        const response = await (
+          await fetch(firebaseEndpoint, {
```

Comment:
You may not need to await for it. Pretty sure that there is a NPM module which provides a bulk API or something like this.

Reply:
Yes and no, currently the few APIs I found are marked as depreciated because Google has its own. And I through we would not like to have the whole firebase library in our code, what do you think?

Reply:
I don't remember if I let the Fixme comment or not about the fact that we need a proper service to handle it with bulk mode someday, on that part I totally agree +1

Reply:
Move it to platform service